src/policy/satisfy.rs

```diff
@@ -0,0 +1,143 @@
+use crate::core::Value;
+use crate::jet::elements::ElementsEnv;
+use crate::policy::ast::Policy;
+use bitcoin_hashes::{sha256, Hash};
+use elements::locktime::Height;
+use elements::schnorr::{KeyPair, XOnlyPublicKey};
+use elements::secp256k1_zkp::Message;
+use elements::taproot::TapLeafHash;
+use elements::SchnorrSig;
+use elements::{LockTime, SchnorrSigHashType, Sequence};
+use elements_miniscript::{Preimage32, Satisfier};
+use std::cmp::Ordering;
+use std::collections::{HashMap, VecDeque};
+use std::iter::FromIterator;
+
+pub struct PolicySatisfier {
+    pub preimages: HashMap<sha256::Hash, Preimage32>,
+    pub keys: HashMap<XOnlyPublicKey, KeyPair>,
+    pub env: ElementsEnv,
+}
+
+impl Satisfier<XOnlyPublicKey> for PolicySatisfier {
+    fn lookup_tap_leaf_script_sig(
+        &self,
+        pk: &XOnlyPublicKey,
+        _: &TapLeafHash,
+    ) -> Option<SchnorrSig> {
+        self.keys.get(pk).map(|keypair| {
+            let sighash = self.env.c_tx_env().sighash_all();
+            let msg = Message::from_hashed_data::<sha256::Hash>(&sighash);
+            let sig = keypair.sign_schnorr(msg);
+
+            SchnorrSig {
+                sig,
+                hash_ty: SchnorrSigHashType::All,
+            }
+        })
+    }
+
+    fn lookup_sha256(&self, hash: &sha256::Hash) -> Option<Preimage32> {
+        self.preimages.get(hash).copied()
+    }
+
+    fn check_older(&self, n: Sequence) -> bool {
+        self.env.tx().input[self.env.ix() as usize].sequence >= n
+    }
+
+    fn check_after(&self, n: LockTime) -> bool {
+        let m = LockTime::from(self.env.tx().lock_time);
+        match m.partial_cmp(&n) {
+            Some(Ordering::Less) | Some(Ordering::Equal) => true,
+            _ => false,
+        }
+    }
+}
+
+pub fn satisfy(policy: &Policy<XOnlyPublicKey>, satisfier: &PolicySatisfier) -> Option<Vec<Value>> {
```

Comment:
In 243e7d6920adbcb1235f1b23de36abb792022614:

I think this architecture where there's just a `satisfy` function that takes a `PolicySatisfier` is good for now. In a later PR we definitely want to abstract over different kinds of satisfiers (in fact, we should be able to just take an arbitrary `elements_miniscript::Satisifier` so we don't even need our own trait).

Reply:
I rebased and implemented `satisfy` for arbitrary `Satisfier<XOnlyPublicKey>`. In a later PR, we should transform this into `Satisfier<Pk>`, but that requires `Policy` to be slightly rewritten in terms of `elements_miniscript`.